google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/impl/GoogleRetry.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.googlecloud.pubsub.impl
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.annotation.InternalApi
+import akka.http.scaladsl.HttpExt
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse, StatusCodes}
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.{RestartSource, Sink, Source}
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+@InternalApi
+private[googlecloud] object GoogleRetry {
+
+  def retryingRequestToResponse(
+      http: HttpExt,
+      request: HttpRequest
+  )(implicit mat: ActorMaterializer): Future[HttpResponse] = {
+    // Exponential backoff as specified in the GCS SLA: https://cloud.google.com/storage/sla
+
+    // 1 initial attempt, plus 2^5 exponential requests to get to 32 seconds
+    val remainingAttempts = new AtomicInteger(6)
+    RestartSource
+      .withBackoff(
+        minBackoff = 1.second,
+        maxBackoff = 32.seconds,
+        randomFactor = 0
+      ) { () =>
```

Comment:
I wonder if this could be solved more idiomatically with a `RetryFlow` (which would need be copied in for Akka 2.5 support).

Reply:
I did not write this part, just moved it over here for reuse.

Reply:
anyway, looking into this

Reply:
Also synced both implementations, their files are almost identical

Reply:
@ennru better like this?